api/server/flexvolume.go

```diff
@@ -1,27 +1,155 @@
 package server
 
-type flexVolumeClient struct{}
+import (
+	"errors"
+	"fmt"
 
-func newFlexVolumeClient() *flexVolumeClient {
-	return &flexVolumeClient{}
-}
+	"github.com/libopenstorage/openstorage/pkg/mount"
+	"github.com/libopenstorage/openstorage/volume"
+
+	"go.pedge.io/dlog"
+)
+
+const (
+	osdDriverKey = "osdDriver"
+	volumeIDKey  = "volumeID"
+	pxDriverKey  = "pxd"
+)
+
+var (
+	deviceDriverMap map[string]string
+	// ErrInvalidSpecDriver is returned when incorrect or no OSD driver is specified in spec file
+	ErrInvalidSpecDriver = errors.New("Invalid kubernetes spec file : No OSD driver specified in flexvolume")
+	// ErrInvalidSpecVolumeID is returned when incorrect or no volumeID is specified in spec file
+	ErrInvalidSpecVolumeID = errors.New("Invalid kubernetes spec file : No volumeID specified in flexvolume")
+	// ErrNoMountInfo is returned when flexvolume is unable to read from proc/self/mountinfo
+	ErrNoMountInfo = errors.New("Could not read mountpoints from /proc/self/mountinfo. ")
+	// ErrMissingMountPath is returned when no mountPath specified in mount call
+	ErrMissingMountPath = errors.New("Missing target mount path")
+)
+
+type flexVolumeClient struct{}
 
 func (c *flexVolumeClient) Init() error {
 	return nil
 }
 
 func (c *flexVolumeClient) Attach(jsonOptions map[string]string) error {
+	driverName, ok := jsonOptions[osdDriverKey]
+	if !ok {
+		return ErrInvalidSpecDriver
+	}
+
+	driver, err := c.getVolumeDriver(driverName)
+	if err != nil {
+		return err
+	}
+
+	mountDevice, ok := jsonOptions[volumeIDKey]
+	if !ok {
+		return ErrInvalidSpecVolumeID
+	}
+
+	if _, err = driver.Attach(mountDevice); err != nil {
+		return err
+	}
 	return nil
 }
 
 func (c *flexVolumeClient) Detach(mountDevice string) error {
+	driverName, ok := deviceDriverMap[mountDevice]
+	if !ok {
+		dlog.Infof("Could not find driver for (%v). Defaulting to pxd", mountDevice)
+		driverName = pxDriverKey
+	}
+
+	driver, err := c.getVolumeDriver(driverName)
+	if err != nil {
+		return err
+	}
+
+	if err = driver.Detach(mountDevice); err != nil {
```

Comment:
same comment about making `err` actually scoped to the if statement, `:=` instead of `=` for the rest of the changes